src/driver_wakeword.cpp

```diff
@@ -14,106 +14,124 @@
-      utt_started = FALSE;
-      E_INFO("Ready....\n");
-    }
-    sleep_msec(10);
-  }
-  ad_close(ad);
+void WakeWordDriver::loadParameters(WakeWordParams wakeword_params) {
+  channel = static_cast<int16_t>(wakeword_params.channel());
+  wakeword = std::string("" + wakeword_params.wake_word());
+  lm_path = std::string("" + wakeword_params.lm_path());
+  dic_path = std::string("" + wakeword_params.dic_path());
+  std::cout << "==> wakeword: " << wakeword << std::endl;
+  std::cout << "==> lenguaje path: " << lm_path << std::endl;
+  std::cout << "==> dictionary path: " << dic_path << std::endl;
 }
 
-} // namespace
+bool WakeWordDriver::startPipe() {
+  std::string cmd = std::string(
+      "malos_psphinx -keyphrase \"" + wakeword +
+      "\" -kws_threshold 1e-20 -dict \"" + dic_path + "\" -lm \"" + lm_path +
+      "\" -inmic yes -adcdev mic_channel" + std::to_string(channel));
 
-namespace matrix_malos {
+  if (!verbose)
+    cmd = cmd + " 2> /dev/null";
 
-bool WakeWordDriver::ProcessConfig(const DriverConfig& config) {
-  // TODO @hpsaturn: Validate all the data that comes from the protos
-  WakeWordParams wakeword_params(config.wakeword());
+  std::cout << "Starting PocketSphinx thread.." << std::endl;
+  if (!(sphinx_pipe_ = popen(cmd.c_str(), "r"))) {
+    return false;
+  }
+  // alsa thread.
+  std::thread pocketsphinx_thread(&WakeWordDriver::PocketSphinxProcess, this);
+  pocketsphinx_thread.detach();
+  returnMatch("voice recognition ready");
 
   return true;
 }
 
-bool WakeWordDriver::SendUpdate() {
+bool WakeWordDriver::stopPipe() {
+  if (sphinx_pipe_ != NULL && enable) {
+    enable = false;
+    std::cout << "Stoping PocketSphinx thread.." << std::endl;
+    if (system(std::string("killall malos_psphinx").c_str()) == -1) {
+      return false;
+    }
+    sleep(2);
+  }
+  return true;
+}
 
+void WakeWordDriver::PocketSphinxProcess() {
+  char buff[512];
+  char match[100];
+  snprintf(match, sizeof(match), "match: %s", wakeword.c_str());
+
+  // processing pipe output
+  while (fgets(buff, sizeof(buff), sphinx_pipe_) != NULL && enable) {
+    std::string sbuff = buff;
+    std::size_t found = sbuff.find(match);
+    if (found != std::string::npos) {
+      std::cout << "[SEND] recognition " << sbuff;
+      sbuff = sbuff.substr(sbuff.find_first_of(" \t") + 1);
+      sbuff = sbuff.erase(sbuff.length() - 1);
+      returnMatch(sbuff);
+    } else {
+      std::cout << "[SKIP] recognition " << sbuff;
+    }
+  }
+  pclose(sphinx_pipe_);
+}
+
+void WakeWordDriver::returnMatch(std::string match) {
+  WakeWordParams wakewordUpdate;
+  wakewordUpdate.set_wake_word(match);
+  std::string buffer;
+  wakewordUpdate.SerializeToString(&buffer);
+  zqm_push_update_->Send(buffer);
+}
+
+bool WakeWordDriver::validatePaths() {
+  if (!exists_path(dic_path))
+    return false;
+  if (!exists_path(lm_path))
+    return false;
   return true;
 }
+
+inline bool WakeWordDriver::exists_path(const std::string &name) {
```

Comment:
You could define this function in the anonymous namespace, above. No need to make it a class member.